File: bzip2_rs_gpt_finished_first_version/src/bzlib/scc_75_BZ2_bzCompress_76_BZ2_bzWriteClose64.rs
// use crate::bzlib::scc_75_BZ2_bzCompress::*;
// use crate::bzlib::scc_41_root_28_prepare_new_block::*;
// use crate::global_vars::bzip2_c1::{blockSize100k, verbosity, workFactor, bz_stream, EState};
use crate::bzlib::scc_40_root_27_prepare_new_block_init_RL_isempty_RL_flush_RL_bzlib_static::*;
use crate::bzlib::scc_74_handle_compress::*;
use crate::global_vars::bzip2_c1::{bz_stream, EState};
use crate::bzlib::bz_constants::*;
pub fn bz2_compress(strm: &mut bz_stream, action: i32) -> i32 {
    // println!("进入 bz2_compress 函数");

    let s = unsafe { &mut *strm.state };
    // println!("75_bzCompress_获取state成功");
    // 请记住，当你将 strm 包裹在 Some 中时
    // 你需要确保 strm 是一个指针类型，
    // 因为 s.strm 是一个 Option<*mut bz_stream> 类型。
    // if s == NULL || s.strm != Some(strm) {
    //     return BZ_PARAM_ERROR;
    // }
    // if s.strm != Some(strm) {
    //     // println!("参数错误：s.strm != Some(strm)");
    //     return BZ_PARAM_ERROR;
    // }

    // println!("75_bzCompress尝试打印打印");
    // print_estate(s, "75_estate_rust.json");
    // println!("状态打印成功");
    let mut counter_75 = 0;
    loop {
        
        // println!("压缩状态：{}", s.mode);
        match s.mode {
            BZ_M_IDLE => {
                // println!("错误：BZ_SEQUENCE_ERROR");
                return BZ_SEQUENCE_ERROR;
            },
            BZ_M_RUNNING => match action {
                BZ_RUN => {
                    // println!("处理 BZ_RUN");
                    let progress = handle_compress(strm);
                    // println!("处理结果：{}", if progress { "成功" } else { "失败" });
                    return if progress { BZ_RUN_OK } else { BZ_PARAM_ERROR };
                },
                BZ_FLUSH => {
                    // println!("处理 BZ_FLUSH");
                    s.avail_in_expect = strm.avail_in;
                    s.mode = BZ_M_FLUSHING;
                    continue;
                },
                BZ_FINISH => {
                    // println!("处理 BZ_FINISH");
                    s.avail_in_expect = strm.avail_in;
                    s.mode = BZ_M_FINISHING;
                    continue;
                },
                _ => {
                    // println!("错误：BZ_PARAM_ERROR");
                    return BZ_PARAM_ERROR;
                },
            },
            BZ_M_FLUSHING => {
                // println!("处于 BZ_M_FLUSHING 状态");
                if action != BZ_FLUSH || s.avail_in_expect != strm.avail_in {
                    // println!("错误：BZ_SEQUENCE_ERROR");
                    return BZ_SEQUENCE_ERROR;
                }
                let progress = handle_compress(strm);
                // println!("处理结果：{}", if progress { "成功" } else { "失败" });
                if s.avail_in_expect > 0 || !is_empty_rl(s) || s.state_out_pos < s.numZ {
                    // println!("返回：BZ_FLUSH_OK");
                    return BZ_FLUSH_OK;
                }
                s.mode = BZ_M_RUNNING;
                // println!("更改状态为 BZ_M_RUNNING");
                return BZ_RUN_OK;
            },
            BZ_M_FINISHING => {
                // println!("处于 BZ_M_FINISHING 状态");
                if action != BZ_FINISH || s.avail_in_expect != strm.avail_in {
                    // println!("错误：BZ_SEQUENCE_ERROR");
                    return BZ_SEQUENCE_ERROR;
                }
                counter_75 += 1;
                // println!("counter_75: {}", counter_75);
                let progress = handle_compress(strm);
                // println!("处理结果：{}", if progress { "成功" } else { "失败" });
                if s.avail_in_expect > 0 || !is_empty_rl(s) || s.state_out_pos < s.numZ {
                    // println!("返回：BZ_FINISH_OK");
                    return BZ_FINISH_OK;
                }
                s.mode = BZ_M_IDLE;
                // println!("更改状态为 BZ_M_IDLE");
                return BZ_STREAM_END;
            },
            _ => {
                // println!("返回：BZ_OK");
                return BZ_OK;
            },
        }
    }
}



// scc_76
use crate::bzlib::scc_24_root_16_BZ2_bzCompressEnd::*;
use std::io::{Error, Write};
use std::ffi::c_void;
use std::ptr;
use crate::global_vars::bzip2_c1::*;
use libc::{ferror, free, FILE};
use std::fs::File;


pub fn bz2_bz_write_close64(
    bzerror: &mut Option<i32>,
    b: BZFILE,
    abandon: i32,
    nbytes_in_lo32: &mut Option<u32>,
    nbytes_in_hi32: &mut Option<u32>,
    nbytes_out_lo32: &mut Option<u32>,
    nbytes_out_hi32: &mut Option<u32>,
) {
    if let Some(bzf) = unsafe { b.as_mut() } {
        if bzf.handle.is_null() { // 修改此处
            set_error(bzerror, None, 0);
            return;
        }
        if !bzf.writing {
            set_error(bzerror, Some(bzf), -1);
            return;
        }
        if unsafe { ferror(bzf.handle as *mut FILE) != 0 } { // 修改此处
            set_error(bzerror, Some(bzf), -6);
            return;
        }

        // 初始化 nbytes_* 参数
        *nbytes_in_lo32 = Some(0);
        *nbytes_in_hi32 = Some(0);
        *nbytes_out_lo32 = Some(0);
        *nbytes_out_hi32 = Some(0);

        if (!abandon != 0) && bzf.lastErr == 0 {
            loop {
                bzf.strm.avail_out = 5000;
                bzf.strm.next_out = bzf.buf.as_mut_ptr() as *mut i8;
                let ret = bz2_compress(&mut bzf.strm, 2);
                if ret != 3 && ret != 4 {
                    set_error(bzerror, Some(bzf), ret);
                    return;
                }

                if bzf.strm.avail_out < 5000 {
                    let n = 5000 - bzf.strm.avail_out as usize;
                    // 安全写入
                    unsafe {
                        if bzf.handle.as_mut().unwrap().write(&bzf.buf[..n]).is_err() {
                            set_error(bzerror, Some(bzf), -6);
                            return;
                        }
                    }
                }

                if ret == 4 {
                    break;
                }
            }
        }

        // 更多的处理...
        // 这里需要根据具体情况补充代码

        // 更新 nbytes_* 参数
        *nbytes_in_lo32 = Some(bzf.strm.total_in_lo32);
        *nbytes_in_hi32 = Some(bzf.strm.total_in_hi32);
        *nbytes_out_lo32 = Some(bzf.strm.total_out_lo32);
        *nbytes_out_hi32 = Some(bzf.strm.total_out_hi32);

        // 清理和结束压缩
        BZ2_bzCompressEnd(&mut Some(bzf.strm));
        unsafe { free(bzf.handle as *mut c_void) };
    } else {
        set_error(bzerror, None, -2);
    }
}

fn set_error(bzerror: &mut Option<i32>, bzf: Option<&mut bzFile>, err: i32) {
    if let Some(error) = bzerror {
        *error = err;
    }
    if let Some(bz_file) = bzf {
        bz_file.lastErr = err;
    }
}

// 你可能需要添加额外的辅助函数和类型

File: bzip2_rs_gpt_finished/src/bzlib/unit_test/scc_75_BZ2_bzCompress_76_BZ2_bzWriteClose64.c

#include "../typedef.h"
#include "./scc_24_root_16_BZ2_bzCompressEnd.c"
#include "./scc_40_root_27_isempty_RL.c"
#include "./scc_74_handle_compress.c"
#include <CUnit/CUnit.h>
#include <CUnit/Basic.h>
// BZ2_bzCompress函数压缩输入数据。
int BZ2_bzCompress(bz_stream *strm, int action) {
   Bool progress;            // 标记函数是否成功执行
   EState* s;

   // 检查strm是否为null
   if (strm == NULL) return (-2);

   s = strm->state;

   // 检查s或strm的状态是否为null
   if (s == NULL || s->strm != strm) return (-2);

   preswitch:
   switch (s->mode) {
      case 1:   // 模式1: 错误
         return (-1);

      case 2:   // 模式2: 压缩数据
         if (action == 0) {
            progress = handle_compress(strm);
            return progress ? 1 : (-2);
         } else if (action == 1) {
            s->avail_in_expect = strm->avail_in;
            s->mode = 3;
            goto preswitch;  // 重新进入switch-case
         } else if (action == 2) {
            s->avail_in_expect = strm->avail_in;
            s->mode = 4;
            goto preswitch;
         } else {
            return (-2);
         }

      case 3:  // 模式3: 继续压缩数据
         if (action != 1 || s->avail_in_expect != s->strm->avail_in)
            return (-1);
         progress = handle_compress(strm);
         if (s->avail_in_expect > 0 || !isempty_RL(s) ||
             s->state_out_pos < s->numZ) return 2;
         s->mode = 2;
         return 1;

      case 4:  // 模式4: 完成压缩
         if (action != 2 || s->avail_in_expect != s->strm->avail_in)
            return (-1);
         progress = handle_compress(strm);
         if (!progress) return (-1);
         if (s->avail_in_expect > 0 || !isempty_RL(s) ||
             s->state_out_pos < s->numZ) return 3;
         s->mode = 1;
         return 4;
   }
   return 0;
}

// 关闭BZIP2文件，但仅提供32位的输入/输出字节计数器
void BZ2_bzWriteClose(int* bzerror, BZFILE* b, int abandon, unsigned int* nbytes_in, unsigned int* nbytes_out) {
    // 使用64位版本的接口，但为64位的计数器传递null指针
    BZ2_bzWriteClose64(bzerror, b, abandon, nbytes_in, NULL, nbytes_out, NULL);
}

// 关闭BZIP2文件，并提供64位的输入/输出字节计数器
void BZ2_bzWriteClose64(
   int* bzerror, BZFILE* b, int abandon,
   unsigned int* nbytes_in_lo32, unsigned int* nbytes_in_hi32,
   unsigned int* nbytes_out_lo32, unsigned int* nbytes_out_hi32) {
   
   Int32 n, n2, ret;
   bzFile* bzf = (bzFile*)b;

   // 检查bzf是否为null
   if (bzf == NULL) {
      if (bzerror) *bzerror = 0;
      return;
   }

   // 检查是否处于写模式
   if (!bzf->writing) {
      if (bzerror) *bzerror = (-1);
      return;
   }

   // 检查文件操作是否有误
   if (ferror(bzf->handle)) {
      if (bzerror) *bzerror = (-6);
      return;
   }

   // 初始化字节计数器为0
   if (nbytes_in_lo32) *nbytes_in_lo32 = 0;
   if (nbytes_in_hi32) *nbytes_in_hi32 = 0;
   if (nbytes_out_lo32) *nbytes_out_lo32 = 0;
   if (nbytes_out_hi32) *nbytes_out_hi32 = 0;

   // 如果没有放弃且上次没有错误，则继续压缩
   if (!abandon && bzf->lastErr == 0) {
      while (1) {
         bzf->strm.avail_out = 5000;
         bzf->strm.next_out = bzf->buf;
         ret = BZ2_bzCompress(&(bzf->strm), 2);
         if (ret != 3 && ret != 4) {
            if (bzerror) *bzerror = ret;
            return;
         }

         if (bzf->strm.avail_out < 5000) {
            n = 5000 - bzf->strm.avail_out;
            n2 = fwrite((void*)(bzf->buf), sizeof(UChar), n, bzf->handle);
            if (n != n2 || ferror(bzf->handle)) {
               if (bzerror) *bzerror = (-6);
               return;
            }
         }

         if (ret == 4) break;
      }
   }

   // 刷新文件缓存
   if (!abandon && !ferror(bzf->handle)) {
      fflush(bzf->handle);
      if (ferror(bzf->handle)) {
         if (bzerror) *bzerror = (-6);
         return;
      }
   }

   // 更新字节计数器的值
   if (nbytes_in_lo32) *nbytes_in_lo32 = bzf->strm.total_in_lo32;
   if (nbytes_in_hi32) *nbytes_in_hi32 = bzf->strm.total_in_hi32;
   if (nbytes_out_lo32) *nbytes_out_lo32 = bzf->strm.total_out_lo32;
   if (nbytes_out_hi32) *nbytes_out_hi32 = bzf->strm.total_out_hi32;

   // 清除错误并结束压缩
   if (bzerror) *bzerror = 0;
   BZ2_bzCompressEnd(&(bzf->strm));
   free(bzf);
}
